### ISSUE TITLE
Performance Issues since TI6 Battle Pass

### ISSUE THREAD

**Issue by iprobie** [ash47/LegendsOfDota#307]
Not sure if everyone here experiences it, but ever since the TI6 battle pass is introduced, there's been frequent performance issues with the custom game.

People getting timed out and crashing all the time, having constant freezes and fps losses.
So far, I've experienced the crashing pretty often and the error code says something about Allocated Memory.

Not sure if this is thanks to gabe newell.

**Comment by darklordabc** [comment]
I've noticed this well. Ive also noticed that if you go to "unassigned" and back to your team many times, dota 2 will eventually crash.

**Comment by ash47** [comment]
I don't think I've played in months. Don't know what I can do. How are
other mods holding up?

**Comment by darklordabc** [comment]
I checked the discussions on a few other mods, and it seems to be only LoD. Prehaps there is some memory leak?
![image](https://cloud.githubusercontent.com/assets/16277198/15425060/7c8de270-1ec9-11e6-88d8-5f0743f4e95f.png)

**Comment by darklordabc** [comment]
This developer had a similair problem: http://steamcommunity.com/workshop/filedetails/discussion/469311655/364040797993532934/
His idea was that the UI was putting too much strain out. That does seem to match up with LoD's problem, as it only seems to crash during the startup where the custom UI is most.

**Comment by darklordabc** [comment]
Roshpit Champions 2.4 is getting smashed by the same problem.

**Comment by iprobie** [comment]
the only other custom game I've played is Overthrow and I've had this issue with overthrow too but not as often as LoD (probably coz I play LoD more, but what the heck).

I also crash sometimes even in-game and not just during the picking/banning phase.

**Comment by ash47** [comment]
TODO: Look into this.

**Comment by Myrl** [comment]
Going to state the obvious, but the fact that it says V8, which I assume is the JS engine, this really seems most likely to be a UI issue.

**Comment by ash47** [comment]
I don't know what I can do :/

**Comment by Myrl** [comment]
Me neither. I'm not an expert of JS myself and I'm not quite sure how one's able to (undeliberately) leak a GC'd language. Maybe this is a weak/strong reference problem or something that I don't know of.

**Comment by darklordabc** [comment]
This is still a major issue. Just played a game then, had at least 2 disconnects during the game, and I had to restart dota 2 exe.

**Comment by darklordabc** [comment]
Can confirm it is something to do wth GUI and ever increasing memory usage. In this video I rapidly change teams in the option select phase, this causes memory usage to grow until the game eventually crashes. I know this does the same thing as shuffling team.
https://drive.google.com/file/d/0B7dpugx95_vadXVpMjgwVkVSeUk/view?usp=drivesdk

**Comment by ash47** [comment]
I roughly know why this occurs, will take a look at patching later today

**Comment by ash47** [comment]
mmm, there are a few snippets of code I have identified that are likely the cause of this issue, I will try to find some time to rewrite them, it's critical code, and needs to be rewritten very precisely otherwise the issues may remain.

**Comment by ash47** [comment]
I have pushed a commit that rewrites most of the code that I THINK is causing the memory leaks -- Please take a look and let me know if you can still reproduce the memory leaking issues?

**Comment by ash47** [comment]
@darklordabc

**Comment by ash47** [comment]
I have a few other ideas on how to fix it, if it is still crashing / leaking

**Comment by darklordabc** [comment]
I'll check it out now.

**Comment by darklordabc** [comment]
There still seems to be an aribtary build up of memory usage during the GUI phase of the game. Interestingly enough, once the game starts, the memory usage remains mostly static. For example, if I rush through the options, and start the game when the memory usage is only around 1,200, it will stay within that number, for the next 5 minutes at least. But if I stay in the menu phase, the memory will continue to increase, especially if I do things like change teams or do actions that make the gui change appearance.

**Comment by darklordabc** [comment]
I'll make a video showing what I mean.

**Comment by darklordabc** [comment]
Hmm, after testing, it seems it has fixed the memory leak issue. I published a test client to the steam workshop (which I will delete), because I wanted to get as close to dedicated server settings as possible. But from this local version, the rapid team switching does not seem to cause the same issues as it did in the previous version. Maybe its fixed? But I'm not sure. 
Below is an annotated video. The first run through i go through GUI phase, as quick as possible, second run through I team switch as much as possible. I was expecting the rapid team switching to eat through memory, but it didn't, and when I started the game, it was using even less memory than the first run-through, strange.
https://drive.google.com/file/d/0B7dpugx95_vabEtVVzZPY0h6d2M/view

**Comment by ash47** [comment]
I basically went through all the code and removed the lazy functions that
generate new panels, and made it cache and reuse them

**Comment by darklordabc** [comment]
I did some more testing. I uploaded an older version of LoD before these memory fixes, and I did the same team switching, and the memory shot up and it crashed before I hit the 3 minutes. I then tried again with the fixed version on the steamworkshop, and it continued to work without the memory leak.

So I think its working. Well at least for the portraits which I think was the main issue, prehaps there is other cases of portraits being created when they could be reused, I dont know. But I think its such an improvement that I would consider pushing an update soon, if I were you. 
Btw, I'll keep the two workshop test files up for the next day, just incase I need to test again, and then I'll delete them.
They are here:
New: http://steamcommunity.com/sharedfiles/filedetails/?id=698879347
Old with memory leak: http://steamcommunity.com/sharedfiles/filedetails/?id=698909518
